### PR TITLE
Added chronological skill/project managing to cli, updated skill date extraction logic. 

### PR DESCRIPTION
**Pull request by 6s-1**
## 📝 Description

This PR enhances the chronological management system by adding CLI tools for editing project and skill dates alongside automatic Git history-based skill date inference. These features work together to provide both automated and manual chronological data management.

Note on pr length: 
These features are tightly coupled and must be deployed together:
-  Database Schema Dependencies: The CLI reads/writes skill dates that the inference logic populates. Without the inference logic, the CLI would only show empty dates. Without the CLI, users can't manually correct inaccurate dates.
-  User Workflow: The complete workflow requires both:
-> Automated inference during project analysis (reduces manual work)
-> Manual corrections via CLI for edge cases (Git bulk imports, non-code skills like "backend"/"frontend")
-  Testing Requirements: Tests cover both the inference logic and the storage layer together, ensuring dates flow correctly from Git → Database → CLI display.

Splitting this PR would result in:
- Non-functional CLI (no dates to display)
- Inference logic with no way to verify or correct results
- Incomplete test coverage
- Broken user experience

**Closes:** #646 #647 #648 

---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [x] ♻️ Refactoring
- [x] ⚡ Performance improvement

---

### Manual Testing - CLI Workflow
 - View chronological data: CLI displays all projects with date ranges
 - Edit project dates: Successfully updated project creation and modification dates
 - View skills by project: Skills display in chronological order with dates
 - Update skill dates: Manually corrected dates for individual skills
 - Remove skills: Successfully removed skills from projects
 - Date validation: Invalid date formats are rejected with helpful error messages

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [x] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

> If applicable, add screenshots to help explain your changes

All tests pass: 
[
Screenshot 2026-02-12 at 8 32 15 PM
](url)

Screenshots to help explain cli flow: 

Screenshot 2026-02-12 at 8 33 05 PM

Screenshot 2026-02-12 at 8 33 19 PM

Screenshot 2026-02-12 at 8 46 04 PM

### CODE REVIEW

File: app/utils/analysis_merger_utils.py [hunk 1]
```diff
@@ -2,13 +2,74 @@
...
 from app.utils.non_code_analysis.non_code_analysis_utils import _sumy_lsa_summarize
 from app.data.db import get_connection
 from app.utils.project_score import compute_overall_project_contribution_score
+from app.utils.git_utils import detect_git, get_repo, is_repo_empty
 MAX_SKILLS = 10 #Maximum number of skills to be stored per project (TDB: adjust based on some condition)
 MAX_BULLETS = 5 #Maximum number of resume bullets to be stored per project (TBD: adjust based on some condition)
 MAX_SENTENCES = 5 #Maximum number of sentences in summary (TBD: adjust based on some condition)
 
+# Skill to file extension mapping for date inference
+SKILL_TO_EXTENSIONS = {
```
Comment (dabby04) [inline app/utils/analysis_merger_utils.py]: Just wanted to confirm, using the skills_to_extensions, how does this differ from what was existing previously in the code_analysis pipeline?

We already had skills defined there, so I want to make sure this isn’t duplicating previous functionality or introducing overlapping logic.

**Review by dabby04 — COMMENTED** [review]
I like the addition of checking for skills based on the first commit. Thank you for the comprehensive description with screenshots. Great work.

Maybe for the main flow, before the user uploads a file, we could ask them if they want to make corrections. It would only show that if they had previously generated insights for their project.

File: app/utils/analysis_merger_utils.py (hunk 1 above)
Comment (6s-1) [reply app/utils/analysis_merger_utils.py]: The system could identify what skills you used (Python, React, SQL) from analyzing your code, but it had no way to determine WHEN you first started using each skill IN Analysis Merger. 

Why It Doesn't Duplicate Anything:
Code analysis pipeline = Identifies WHAT skills exist (content analysis)
SKILL_TO_EXTENSIONS = Determines WHEN skills were first used in both pipelines

**Comment by 6s-1** [comment]
> I like the addition of checking for skills based on the first commit. Thank you for the comprehensive description with screenshots. Great work.
> 
> Maybe for the main flow, before the user uploads a file, we could ask them if they want to make corrections. It would only show that if they had previously generated insights for their project.

Thank you for the suggestion! I have implemented the same for when previously analysed insights are detected.

**Comment by 6s-1** [comment]
> The changes look good and thank you for the explanation. All tests pass.
> 
> Image
> Could we possibly move the corrections statement above the project analysis session since it doesn't fall under analysis.

Sure, I have implemented the same!

File: app/utils/analysis_merger_utils.py [hunk 2]
```diff
@@ -270,6 +331,124 @@ def get_project_score(code_metrics, non_code_metrics, git_metrics):
...
+    
+    # Try partial match (skill contains or is contained in)
+    skill_lower = skill.lower()
+    for skill_name, extensions in SKILL_TO_EXTENSIONS.items():
+        if skill_lower in skill_name.lower() or skill_name.lower() in skill_lower:
+            return extensions
+    
+    return []
+
+def _infer_skill_dates_from_git(project_path: str, skills: List[str]) -> Dict[str, Optional[str]]:
```
Comment (kjassani) [inline app/utils/analysis_merger_utils.py]: It looks like`_infer_skill_dates_from_git()` currently does not filter commits by the target user. 

It scans all repo commits and assigns the earliest date where a matching file extension appears. That would mean the stored skill date may reflect when any contributor first used that technology, not when the current app user first used it. I thought that skill chronology is intended to be user-specific, and this should filter commits with author matching before selecting earliest dates. Please let me know if our understanding of the requirement and implementation is correct.

Comment (6s-1) [reply app/utils/analysis_merger_utils.py]: That's a good catch Karim, I've implemented user-specific commit filtering, This ensures skill dates reflect when the user first used each technology, not when any team member did.

File: app/utils/analysis_merger_utils.py [hunk 3]
```diff
@@ -270,6 +332,134 @@ def get_project_score(code_metrics, non_code_metrics, git_metrics):
...
+                skill_dates[skill] = None
+                continue
+            
+            earliest_date = None
+            
+            # Iterate through all commits (oldest first)
+            try:
+                for commit in repo.iter_commits(rev="--all", reverse=True):
+                    # Filter commits by user email if available
+                    if user_email and commit.author.email != user_email:
```
[the remark this answers is not in the document]
Comment (6s-1) [reply app/utils/analysis_merger_utils.py]: Thank you for pointing that out Karim, the old logic took into account the author match but not necessarily the no reply email issue which I have now implemented.
